### PR TITLE
Sum type for JSON and YAML simple object notations

### PR DESCRIPTION
**Pull request by tjanc**
- `std::variant`-like tagged union type with unit tests
- added `boost::containers`, enabling recursive data structures and opening the door to `boost::pmr`
- definitions and aliases for JSON/YAML values
- outline for JSON serialization

### CODE REVIEW

File: test/utils/test-Variant.cc [hunk 1]
```diff
@@ -0,0 +1,895 @@
...
+            {
+                REQUIRE(Foo::record().copy_constructor == 0);
+                REQUIRE(Foo::record().copy_assignment == 0);
+            }
+
+            THEN("it was not moved")
+            {
+                REQUIRE(Foo::record().copy_constructor == 0);
+                REQUIRE(Foo::record().copy_assignment == 0);
+            }
```
Comment (pksunkara) [inline test/utils/test-Variant.cc]: I don't see the above 2 `THEN` anywhere else in similar tests below. Did you forget to remove these?

File: test/utils/test-Variant.cc [hunk 2]
```diff
@@ -0,0 +1,895 @@
...
+        Bar::reset_record();
+        Baz::reset_record();
+
+        tested v(std::move(bar));
+
+        THEN("one Bar and one Foo exist")
+        {
+            REQUIRE(Foo::instances().size() == 0);
+            REQUIRE(Bar::instances().size() == 2);
+            REQUIRE(Baz::instances().size() == 0);
```
Comment (pksunkara) [inline test/utils/test-Variant.cc]: The tests and the message is contradicting.

File: test/utils/test-Variant.cc (hunk 1 above)
Comment (tjanc) [reply test/utils/test-Variant.cc]: 👍 actually `move_contructor` and `move_assignment` should be tested there

File: test/utils/test-Variant.cc (hunk 2 above)
Comment (tjanc) [reply test/utils/test-Variant.cc]: :+1 wrongly copy-pasted message

File: test/utils/test-Variant.cc (hunk 1 above)
Comment (pksunkara) [reply test/utils/test-Variant.cc]: Shouldn't you be testing `move_assignment` and `copy_assignment` in other tests below too?

Comment (tjanc) [reply test/utils/test-Variant.cc]: 👍 added assertions where missing

**Comment by klokane** [comment]
What is source `test/fixtures/utf-8` there is missing license or something like this.

**Comment by klokane** [comment]
Missing tests for variant `visit`, `emplace`

**Comment by klokane** [comment]
Do we need `boost` libs at this time?

**Comment by klokane** [comment]
Missing tests for src/utils/Utils.h

**Comment by tjanc** [comment]
- `test/fixtures/utf-8` is not added as part of this PR but #527 ; however, it's just a sequence of all utf-8 characters, so I don't really think licensing is an issue
- tests for visit are missing and should definitely be included; emplace is already tested by testing `operator=` but I can copy-paste the tests
- we need boost to represent recursive data structures; boost containers specifically allow us to do that; that's also in the description of this PR
- `static_assert`s for `src/utils/Utils.h` are missing and should be included

**Comment by tjanc** [comment]
I will fix this in a separate PR, ok?

**Comment by klokane** [comment]
It is ok fix it other PR.
About `test/fixtures/utf-8` I mean if it come from someone other (you did not create it by yourself) then you will need some kind of license

**Pull request merged**